crates/rover-client/src/operations/persisted_queries/publish/types.rs

```diff
@@ -107,6 +118,71 @@ impl From<PersistedQueriesPublishInput> for QueryVariables {
     }
 }
 
+#[derive(Debug, Serialize, Deserialize)]
+pub struct RelayPersistedQueryManifest {
+    #[serde(flatten)]
+    operations: HashMap<String, String>,
+}
+
+impl TryFrom<RelayPersistedQueryManifest> for ApolloPersistedQueryManifest {
+    type Error = RoverClientError;
+
+    fn try_from(relay_manifest: RelayPersistedQueryManifest) -> Result<Self, Self::Error> {
+        let mut compiler = ApolloCompiler::new();
+
+        let mut diagnostics = Vec::new();
+        let mut operations = Vec::new();
+        let mut id_mapper = HashMap::new();
+        for (id, body) in &relay_manifest.operations {
+            let file_id = compiler.add_document(body, id);
+            id_mapper.insert(file_id, (id, body));
+        }
+        for operation in compiler.db.all_operations().iter() {
+            let file_id = operation.loc().file_id();
+            diagnostics.extend(compiler.db.validate_standalone_executable(file_id));
```

Comment:
While I appreciate that @goto-bus-stop created this function for us, which definitely seems like it could be useful in general, I'm not sure it's the right thing to use here.

My concern is that GraphQL does evolve over time and as things are added to the spec, it's nice if you don't have to obsessively upgrade every single component of the system. We're probably good about keeping Rover up to date but we certainly have had problems where parts of GraphOS got behind and didn't support things like deprecated arguments or whatever.

The only reason we need to parse this at all is that we decided that a manifest needs to contain an operation name and an operation type, and Relay doesn't include that information. (The existing case with our manifest format doesn't parse the operations at all, which keeps things a lot less tightly coupled.)

So I think we should do the bare minimum required to determine that the operation contains a unique named operation and pull them out.  I do not think that it's Rover's job to refuse to send a string to GraphOS if it has opinions about "variable usage" or "fragment usage" or whatever. I don't want to be in position where users are blocked from using some fancy new GraphQL capability that Relay Compiler understands and Router understands but some pieces in the middle don't. It's bad enough that syntactic language extensions will have this issue (because we are required to parse the document); I don't think that semantic extensions relating to validations other than "is there a unique named operation?" should cause these issues.